enterprise/app/trends/heatmap.tsx

```diff
@@ -0,0 +1,422 @@
+    if (!width) {
+      return null;
+    }
+    const numColumns = this.props.heatmapData.column.length || 1;
+    const xTickMod = Math.ceil(numColumns / Math.min(numColumns, width / TICK_LABEL_SPACING_MAGIC_NUMBER));
+
+    return (
+      <g color="#666" transform={`translate(${CHART_MARGINS.left}, ${this.props.height - CHART_MARGINS.bottom})`}>
+        <line stroke="#666" x1="0" y1="0" x2={width} y2="0"></line>
+        {this.xScaleBand.domain().map((v, i) => {
+          const tickX = this.xScaleBand.bandwidth() * i;
+
+          return (
+            <g transform={`translate(${this.xScaleBand.bandwidth() * i}, 0)`}>
+              {i % xTickMod == 0 && (
+                <text fill="#666" x={this.xScaleBand.bandwidth() / 2} y="18" fontSize="12" textAnchor="middle">
+                  {moment(+v / 1000).format("MMM D")}
+                </text>
+              )}
+              <line stroke="#666" x1="0" y1="0" x2="0" y2="4"></line>;
+            </g>
+          );
+        })}
+      </g>
+    );
+  }
+
+  renderYAxis(height: number): JSX.Element | null {
+    if (!height) {
+      return null;
+    }
+    const numRows = this.numHeatmapRows();
+    const yTickMod = Math.ceil(numRows / Math.min(numRows, height / TICK_LABEL_SPACING_MAGIC_NUMBER));
+
+    return (
+      <g
+        color="#666"
+        transform={`translate(${CHART_MARGINS.left}, ${this.props.height - CHART_MARGINS.bottom - height})`}>
+        <line stroke="#666" x1="0" y1="0" x2="0" y2={height}></line>
+        {this.yScaleBand.domain().map((v, i) => {
+          return (
+            <g transform={`translate(0, ${height - this.yScaleBand.bandwidth() * i})`}>
+              {i % yTickMod == 0 && (
+                <text fill="#666" x="-8" y="3" fontSize="12" textAnchor="end">
+                  {this.renderYBucketValue(i)}
+                </text>
+              )}
+              <line stroke="#666" x1="0" y1="0" x2="-4" y2="0"></line>
+            </g>
+          );
+        })}
+      </g>
+    );
+  }
+
+  render() {
+    const width = this.props.width - CHART_MARGINS.left - CHART_MARGINS.right;
+    const height = this.props.height - CHART_MARGINS.top - CHART_MARGINS.bottom;
+
+    const xDomain = this.props.heatmapData.timestampBracket.slice(0, -1).map((v) => +v);
+    const yDomain = this.props.heatmapData.bucketBracket
+      .slice(0, -1)
+      .map((v) => +v)
+      .reverse();
+
+    this.xScaleBand = scaleBand<number>().range([0, width]).domain(xDomain);
+    this.yScaleBand = scaleBand<number>().range([0, height]).domain(yDomain);
+
+    let min = 0;
+    let max = 0;
+    this.props.heatmapData.column.map((column) => {
+      column.value.map((value) => {
+        min = Math.min(+value, min);
+        max = Math.max(+value, max);
+      });
+    });
+
+    const selection = this.maybeRenderSelection();
+    const interpolator = (v: number) =>
+      heatmapColorScale[Math.floor((heatmapColorScale.length * 0.99 * (v - min)) / (max - min))];
```

Comment:
nit: the `* 0.99` looks a little weird, is there a cleaner way to do this calculation? (`length - 1`  maybe?)

Reply:
I was trying to avoid getting screwed up by floating point division with my floor() call.  thinking about this a bit harder, i *think* i'm being overly paranoid because i'm not going to be using huge values and v === max should cleanly pop out as integer 1.